# Part IA Computing: Michaelmas Term
## Exercises 01.1 and 01.2

Edwin Bahrami Balani ([`eb677`](mailto:eb677@cam.ac.uk))

> ## Exercise 01.1
> 
> Degrees Fahrenheit ($T_F$) are converted to degrees Celsius ($T_c$) using the formula:
> 
> $$
T_c = 5(T_f - 32)/9
$$
> 
> Write a program to convert 78 degrees Fahrenheit to degrees Celsius and print the result.
> Write your program such that you can easily change the temperature in Fahrenheit that you are converting from.

In [1]:
T_f = 78
T_c = 5 * (T_f - 32) / 9

print(T_c)


25.555555555555557


> ## Exercise 01.2
> 
> Interest on a bank loan is charged at fixed rate above the Bank of England 'official Bank Rate'. 
> 
> 1. Write a program that computes the interest payable, using the following variables in your program:
> 
>     1. Loan principal (amount borrowed)
>     1. Official Bank Rate (percentage, expressed per annum))
>     1. Rate over the official Bank Rate (percentage, expressed per annum)
>     1. Time in days
> 
>   Test your program with a loan principal of £150,000, bank rate of 0.25%, rate over the 
>   bank rate of 1.49% and a period of 28 days.
>     
> 1. Modify your program such that it prompts the user to interactively enter the principal and the 
>    number of days.
> 
> *Hint*: To prompt a user to input numerical data, use the command:
> 
>     # Get input from user
>     number = float(input('How many seconds in one minute? '))
> 
> Do not forget to use `float`. The reason for this will be explained later.

In [2]:
# Part IA Computing: Michaelmas Term
# Edwin Bahrami Balani (eb677@cam.ac.uk)

# Exercise 01.2
# Version A - no user input

principal = 150000
bank_rate = 0.25 / 100
rate_over_br = 1.49 / 100
period = 28

interest_rate = bank_rate + rate_over_br

interest = principal * interest_rate * (period / 365)

def p_tab(*args):
    print(*args, sep='\t')

p_tab("Principal:\t\t", '£ '+str(principal))
p_tab("Official Bank Rate:\t", str(bank_rate*100)+'%')
p_tab("Rate over Official Bank Rate:", str(rate_over_br*100)+'%')
p_tab("Period:\t\t\t", str(period)+" days")
p_tab()
p_tab("Calculated interest rate:", str(interest_rate*100)+'%') # Float arithmetic...

p_tab("INTEREST PAYABLE = \t", '£ '+str(interest))

Principal:			£ 150000
Official Bank Rate:		0.25%
Rate over Official Bank Rate:	1.49%
Period:				28 days

Calculated interest rate:	1.7399999999999998%
INTEREST PAYABLE = 		£ 200.2191780821918


## Explanation of the program
The *effective* interest rate of our loan equals $\text{official bank rate} + \text{rate over bank rate}$

Thus, $\text{interest rate} = 0.25\% + 1.49\% = 1.74\%$

However, while `interest_rate` is per annum, our `period` is expressed in days. We will assume that the interest is not compounded, i.e. interest is only payable on the principal, not on already-accumulated interest as well.

So, the interest payable in a given period of days is simply $\text{interest payable per year} \times \frac{\text{period in days}}{365}$.

$$
\text{interest payable per year} = \text{principal} \times \text{interest rate} \\
\therefore \text{interest payable} = \text{principal} \times \text{interest rate} \times \frac{\text{period in days}}{365}
$$

In [3]:
# Part IA Computing: Michaelmas Term
# Edwin Bahrami Balani (eb677@cam.ac.uk)

# Exercise 01.2
# Version B - some user input

principal = float(input("Enter loan principal: "))
bank_rate = 0.25 / 100
rate_over_br = 1.49 / 100
period = int(input("Enter loan period in days: "))

interest_rate = bank_rate + rate_over_br

interest = principal * interest_rate * (period / 365)

def p_tab(*args):
    print(*args, sep='\t')

p_tab("Principal:\t\t", '£ '+str(principal))
p_tab("Official Bank Rate:\t", str(bank_rate*100)+'%')
p_tab("Rate over Official Bank Rate:", str(rate_over_br*100)+'%')
p_tab("Period:\t\t\t", str(period)+" days")
p_tab()
p_tab("Calculated interest rate:", str(interest_rate*100)+'%') # Float arithmetic...

p_tab("INTEREST PAYABLE = \t", '£ '+str(interest))

Enter loan principal: 1584654
Enter loan period in days: 87
Principal:			£ 1584654.0
Official Bank Rate:		0.25%
Rate over Official Bank Rate:	1.49%
Period:				87 days

Calculated interest rate:	1.7399999999999998%
INTEREST PAYABLE = 		£ 6572.189658082191
